File: tsne_pso/tests/test_tsne_pso.py
import numpy as np
import pytest
import time
from tsne_pso import TSNE_PSO

def test_basic_functionality():
    # Generate random data
    rng = np.random.RandomState(42)
    X = rng.randn(100, 10)
    
    # Initialize t-SNE
    tsne = TSNE_PSO(
        n_components=2,
        perplexity=30.0,
        n_particles=50,
        max_iter=250,
        random_state=42,
        verbose=False
    )
    
    # Fit and transform
    embedding = tsne.fit_transform(X)
    
    # Check output shape
    assert embedding.shape == (100, 2)
    
    # Check that output is not all zeros
    assert not np.allclose(embedding, 0)

def test_different_dimensions():
    rng = np.random.RandomState(42)
    X = rng.randn(50, 20)
    
    for n_components in [2, 3, 5]:
        tsne = TSNE_PSO(n_components=n_components)
        embedding = tsne.fit_transform(X)
        assert embedding.shape == (50, n_components)

def test_perplexity_effect():
    rng = np.random.RandomState(42)
    X = rng.randn(75, 15)
    
    embeddings = []
    perplexities = [5.0, 30.0, 50.0]
    
    for perp in perplexities:
        tsne = TSNE_PSO(perplexity=perp, random_state=42)
        embeddings.append(tsne.fit_transform(X))
    
    # Check that different perplexities give different results
    for i in range(1, len(embeddings)):
        assert not np.allclose(embeddings[i], embeddings[i-1])

def test_random_state():
    X = np.random.randn(50, 10)
    
    # Same random state should give same results
    tsne1 = TSNE_PSO(random_state=42)
    tsne2 = TSNE_PSO(random_state=42)
    
    embedding1 = tsne1.fit_transform(X)
    embedding2 = tsne2.fit_transform(X)
    
    assert np.allclose(embedding1, embedding2)
    
    # Different random states should give different results
    tsne3 = TSNE_PSO(random_state=43)
    embedding3 = tsne3.fit_transform(X)
    
    assert not np.allclose(embedding1, embedding3)

def test_input_validation():
    tsne = TSNE_PSO()
    
    # Test empty input
    with pytest.raises(ValueError):
        tsne.fit_transform(np.array([]))
    
    # Test inconsistent dimensions
    X = np.random.randn(10, 5)
    X = np.vstack([X, np.random.randn(1, 6)])  # Add row with different dimension
    with pytest.raises(ValueError):
        tsne.fit_transform(X)

def test_parallel_execution():
    X = np.random.randn(100, 20)
    
    # Test with different numbers of threads
    times = []
    for n_jobs in [1, 2, 4]:
        tsne = TSNE_PSO(n_jobs=n_jobs, max_iter=100)
        
        start_time = time.time()
        tsne.fit_transform(X)
        times.append(time.time() - start_time)
    
    # More threads should generally be faster
    if len(times) >= 2:
        avg_speedup = np.mean([times[0] / t for t in times[1:]])
        assert avg_speedup > 1.0

def test_version():
    tsne = TSNE_PSO()
    assert isinstance(tsne.version(), str)
    assert len(tsne.version()) > 0 

File: tests/unit/test_tsne_pso.cpp
#include <gtest/gtest.h>
#include "tsne_pso.hpp"
#include <vector>
#include <random>
#include <chrono>

using namespace tsne_pso;

class TSNEPSOTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Generate random test data
        std::mt19937 gen(42);
        std::normal_distribution<double> dist(0.0, 1.0);
        
        n_samples = 100;
        n_features = 10;
        test_data.resize(n_samples, std::vector<double>(n_features));
        
        for (auto& sample : test_data) {
            for (auto& feature : sample) {
                feature = dist(gen);
            }
        }
    }
    
    std::vector<std::vector<double>> test_data;
    size_t n_samples;
    size_t n_features;
};

// Test basic fitting functionality
TEST_F(TSNEPSOTest, BasicFit) {
    TSNE_PSO tsne(2, 30.0, 50, 250, 200.0, 12.0, 42, 1, false);
    
    auto embedding = tsne.fit_transform(test_data);
    
    ASSERT_EQ(embedding.size(), n_samples);
    ASSERT_EQ(embedding[0].size(), 2);
    
    // Check that embedding is not all zeros
    bool all_zeros = true;
    for (const auto& sample : embedding) {
        for (const auto& coord : sample) {
            if (std::abs(coord) > 1e-10) {
                all_zeros = false;
                break;
            }
        }
        if (!all_zeros) break;
    }
    EXPECT_FALSE(all_zeros);
}

// Test that different random seeds produce different results
TEST_F(TSNEPSOTest, RandomSeedEffect) {
    TSNE_PSO tsne1(2, 30.0, 50, 100, 200.0, 12.0, 42);
    TSNE_PSO tsne2(2, 30.0, 50, 100, 200.0, 12.0, 43);
    
    auto embedding1 = tsne1.fit_transform(test_data);
    auto embedding2 = tsne2.fit_transform(test_data);
    
    // Check that results are different
    bool all_same = true;
    for (size_t i = 0; i < n_samples; ++i) {
        for (size_t j = 0; j < 2; ++j) {
            if (std::abs(embedding1[i][j] - embedding2[i][j]) > 1e-10) {
                all_same = false;
                break;
            }
        }
        if (!all_same) break;
    }
    EXPECT_FALSE(all_same);
}

// Test that output dimensions are correct
TEST_F(TSNEPSOTest, OutputDimensions) {
    std::vector<int> test_dimensions = {2, 3, 5};
    
    for (int dim : test_dimensions) {
        TSNE_PSO tsne(dim);
        auto embedding = tsne.fit_transform(test_data);
        
        EXPECT_EQ(embedding.size(), n_samples);
        EXPECT_EQ(embedding[0].size(), dim);
    }
}

// Test behavior with different perplexity values
TEST_F(TSNEPSOTest, PerplexityEffect) {
    std::vector<double> perplexities = {5.0, 30.0, 50.0};
    std::vector<std::vector<std::vector<double>>> embeddings;
    
    for (double perp : perplexities) {
        TSNE_PSO tsne(2, perp);
        embeddings.push_back(tsne.fit_transform(test_data));
    }
    
    // Check that different perplexities give different results
    for (size_t i = 1; i < embeddings.size(); ++i) {
        bool all_same = true;
        for (size_t j = 0; j < n_samples; ++j) {
            for (size_t k = 0; k < 2; ++k) {
                if (std::abs(embeddings[i][j][k] - embeddings[i-1][j][k]) > 1e-10) {
                    all_same = false;
                    break;
                }
            }
            if (!all_same) break;
        }
        EXPECT_FALSE(all_same);
    }
}

// Test parallel execution
TEST_F(TSNEPSOTest, ParallelExecution) {
    std::vector<int> thread_counts = {1, 2, 4};
    std::vector<double> execution_times;
    
    for (int threads : thread_counts) {
        TSNE_PSO tsne(2, 30.0, 50, 100, 200.0, 12.0, 42, threads);
        
        auto start = std::chrono::high_resolution_clock::now();
        tsne.fit_transform(test_data);
        auto end = std::chrono::high_resolution_clock::now();
        
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
        execution_times.push_back(duration);
    }
    
    // Check that more threads generally means faster execution
    if (execution_times.size() >= 2) {
        double avg_speedup = 0.0;
        for (size_t i = 1; i < execution_times.size(); ++i) {
            avg_speedup += execution_times[0] / execution_times[i];
        }
        avg_speedup /= (execution_times.size() - 1);
        
        // Expect some speedup with more threads
        EXPECT_GT(avg_speedup, 1.0);
    }
}

// Test version string
TEST_F(TSNEPSOTest, VersionString) {
    std::string version = TSNE_PSO::version();
    EXPECT_FALSE(version.empty());
} 